cedar-lean/DiffTest/Parser.lean

```diff
@@ -36,296 +36,334 @@ open Cedar.Spec
+def jsonToValue (json : Lean.Json) : ParseResult Value :=
+  jsonToExpr json >>= exprToValue
 
-def jsonToContext (json : Lean.Json) : Map Attr Value :=
-  let value := jsonToValue json
+def jsonToContext (json : Lean.Json) : ParseResult (Map Attr Value) := do
+  let value ← jsonToValue json
   match value with
-  | .record kvs => kvs
-  | _ => panic! "jsonToContext: context must be a record\n" ++ toString (repr value)
+  | .record kvs => .ok kvs
+  | _ => .error ("jsonToContext: context must be a record\n" ++ toString (repr value))
 
 /-
 The "Known" in this function refers to "known" vs. "unknown" entities.
 We only need to support the known case here because the Lean does not
 support partial evaluation.
 -/
-def jsonToRequest (json : Lean.Json) : Request :=
-  let principal := getJsonField (getJsonField json "principal") "Known"
-  let action := getJsonField (getJsonField json "action") "Known"
-  let resource := getJsonField (getJsonField json "resource") "Known"
-  let context := getJsonField json "context"
-  {
-    principal := jsonToEuid principal,
-    action := jsonToEuid action,
-    resource := jsonToEuid resource,
-    context := jsonToContext context
+def jsonToRequest (json : Lean.Json) : ParseResult Request := do
+  let principal ← getJsonField json "principal" >>= (getJsonField · "Known") >>= jsonToEuid
+  let action ← getJsonField json "action" >>= (getJsonField · "Known") >>= jsonToEuid
+  let resource ← getJsonField json "resource" >>= (getJsonField · "Known") >>= jsonToEuid
+  let context ← getJsonField json "context" >>= jsonToContext
+  .ok {
+    principal := principal,
+    action := action,
+    resource := resource,
+    context := context
   }
 
-def jsonToEntityData (json : Lean.Json) : EntityData :=
-  let ancestorsArr := jsonToArray (getJsonField json "ancestors")
-  let ancestors := Set.mk (List.map jsonToEuid ancestorsArr.toList)
-  let attrsKVs := jsonObjToKVList (getJsonField json "attrs")
-  let attrs := Map.mk (List.map (λ (k,v) => (k,jsonToValue v)) attrsKVs)
-  {
-    ancestors := ancestors,
-    attrs := attrs
+def jsonToEntityData (json : Lean.Json) : ParseResult EntityData := do
+  let ancestorsArr ← getJsonField json "ancestors" >>= jsonToArray
+  let ancestors ← List.mapM jsonToEuid ancestorsArr.toList
+  let attrsKVs ← getJsonField json "attrs" >>= jsonObjToKVList
+  let attrs ← List.mapM (λ (k,v) => jsonToValue v >>= λ v => .ok (k,v)) attrsKVs
+  .ok {
+    ancestors := Set.mk ancestors,
+    attrs := Map.mk attrs
   }
 
-def jsonToEntities (json : Lean.Json) : Entities :=
-  let entities := getJsonField json "entities"
-  let kvs := jsonArrayToKVList entities
-  Map.mk (List.map (λ (k,v) => (jsonToEuid k, jsonToEntityData v)) kvs)
+def jsonToEntities (json : Lean.Json) : ParseResult Entities := do
+  let entities ← getJsonField json "entities"
+  let kvs_json ← jsonArrayToKVList entities
+  let kvs ← List.mapM (λ (k,v) => jsonToEuid k >>= λ k => jsonToEntityData v >>= λ v => .ok (k, v)) kvs_json
```

Comment:
instead of this string of `>>= λ`, could we use a nested `do` block, or does that not work in Lean?